core/src/sumeragi/main_loop.rs

```diff
@@ -699,113 +799,86 @@ pub fn run<F>(
-                            && proxy_tail_signatures.len() == 1
-                        {
-                            // TODO: An observing peer should not validate, yet we will do so
-                            // in order to preserve old behaviour. This should be changed.
-                            // Tracking issue : https://github.com/hyperledger/iroha/issues/2635
-                            let block = match block.revalidate(
-                                &sumeragi.transaction_validator,
-                                &state.wsv,
-                                &state.latest_block_hash,
-                                state.latest_block_height,
-                            ) {
-                                Ok(block) => block,
-                                Err(err) => {
-                                    warn!(%err);
-                                    continue;
-                                }
-                            };
-                            for event in Vec::<Event>::from(&block) {
-                                trace!(?event);
-                                sumeragi.events_sender.send(event).unwrap_or(0);
+                        continue;
+                    }
+
+                    let block = {
+                        let span = span!(
+                            Level::TRACE,
+                            "Sumeragi Observing Peer is revalidating the block."
+                        );
+                        let _enter = span.enter();
+                        match block.revalidate(
+                            &sumeragi.transaction_validator,
+                            &state.wsv,
+                            &state.latest_block_hash,
+                            state.latest_block_height,
+                        ) {
+                            Ok(block) => block,
+                            Err(err) => {
+                                warn!(%err);
+                                continue;
                             }
-                            commit_block(sumeragi, block, &mut state);
                         }
+                    };
+
+                    for event in Vec::<Event>::from(&block) {
+                        trace!(?event);
+                        sumeragi.events_sender.send(event).unwrap_or(0);
                     }
-                    _ => {
-                        trace!("Observing peer not handling message {:?}", incoming_message);
-                    }
+
+                    let block_clone = block.clone();
+                    let key_pair_clone = sumeragi.key_pair.clone();
+                    let signed_block = block_clone
+                        .sign(key_pair_clone)
+                        .expect("maybe we should handle this error");
+
+                    let voting_block = VotingBlock::new(signed_block.clone());
+                    voting_block_option = Some(voting_block);
```

Comment:
I presume this is for your bandwidth saving PR? Making sure observing peers have a copy of the block so that the proxy tail message can be changed to only include the hash of the block?

Reply:
yeap